### PR TITLE
edk II C Coding Standard: File and directory naming guidelines

### PR DESCRIPTION
**Pull request by changab**
Add file and directory naming guidelines for EDKII modules. Also consider the processor architecture and vendor implementation.

This is the draft version to community for the wide discussion.

Signed-off-by: Abner Chang <abner.chang@amd.com>
Cc: Ray Ni <ray.ni@intel.com>
Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Sunil V L <sunilvl@ventanamicro.com>
Cc: Abdul Lateef Attar <abdattar@amd.com>

### CODE REVIEW

File: 4_naming_conventions/42_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_file_names.md]: The above is not necessary because it's a special case of the following rule.

File: 4_naming_conventions/42_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_file_names.md]: In MdeModulePkg, PCD driver follows this rule. But ResetSystemRuntimeDxe and ResetSystemPei don't follow.
Can we remove this rule so that the rule is clearer.

File: 4_naming_conventions/42_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_file_names.md]: How about put the directory naming rule before the file naming rule?

File: 4_naming_conventions/42_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_file_names.md]: I do not understand the following rules.

File: 4_naming_conventions/42_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_file_names.md]: Good to me.

File: 4_naming_conventions/42_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_file_names.md]: shall it be moved to directory chapter?

File: 4_naming_conventions/46_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/46_directory_names.md]: again, can we move this chapter before file naming chapter?

File: 4_naming_conventions/46_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/46_directory_names.md]: Shall we say \<Feature\>\<Phase\> so that not introducing another \<ModuleName\> keyword?

File: 4_naming_conventions/46_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/46_directory_names.md]: good to me except replace ModuleName with Feature

File: 4_naming_conventions/46_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/46_directory_names.md]: With the replacement in above rule, this rule is not needed.

File: 4_naming_conventions/46_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/46_directory_names.md]: Good to me.

**Review by niruiyu — CHANGES_REQUESTED** [review]

File: 4_naming_conventions/46_directory_names.md
Comment (vlsunil) [inline 4_naming_conventions/46_directory_names.md]: Wondering whether we can have better example. Smm comes twice.

File: 4_naming_conventions/42_file_names.md
Comment (changab) [reply 4_naming_conventions/42_file_names.md]: This is related to the chapter structure in this spec. 4.2 is mainly for the file, so add Directory Names in 4.6.

File: 4_naming_conventions/42_file_names.md
Comment (changab) [reply 4_naming_conventions/42_file_names.md]: I think the above one says if the module is generic to all CPU archs or >=2 CPU archs (X64 and IA32)? @mdkinney 
The following one should be changed to "If the implementation is considered as only for the specific CPU arch or vendor. 
<CpuArch> should be REQUIRED, right?

File: 4_naming_conventions/42_file_names.md
Comment (changab) [reply 4_naming_conventions/42_file_names.md]: Or the spec is only applied to the newly introduced module or reconstructions after the doc is published?

File: 4_naming_conventions/42_file_names.md
Comment (changab) [reply 4_naming_conventions/42_file_names.md]: This seems the files and directory mixed.

File: 4_naming_conventions/46_directory_names.md
Comment (changab) [reply 4_naming_conventions/46_directory_names.md]: What are the impacts to insert a section in between?

File: 4_naming_conventions/42_file_names.md
Comment (changab) [reply 4_naming_conventions/42_file_names.md]: What are the impacts to insert a section in between?

File: 4_naming_conventions/46_directory_names.md
Comment (changab) [reply 4_naming_conventions/46_directory_names.md]: need more details of this question

File: 4_naming_conventions/46_directory_names.md
Comment (changab) [reply 4_naming_conventions/46_directory_names.md]: The <modulename> is not the same as the <feature> in the usual case?

File: 4_naming_conventions/42_file_names.md
Comment (mdkinney) [reply 4_naming_conventions/42_file_names.md]: No issues with inserting sections within.  Just be aware if you reorder items that are linked in Summary.md, you have to update that file too

File: 4_naming_conventions/42_file_names.md
Comment (mdkinney) [reply 4_naming_conventions/42_file_names.md]: Both styles are allowed.  It is developer choice.

File: 4_naming_conventions/42_file_names.md
Comment (mdkinney) [reply 4_naming_conventions/42_file_names.md]: We have to be careful here.  Just because the very first implementation of a component may be for a single CPU or CPU+Vendor, that does not mean that additional CPU/Vendor support may not be added in the future.  We do not want to change the directory of INF name later because that breaks backwards compatibility.  The rule below must only be used if the developer can not imagine any other CPU/Vendor combinations every being added later

File: 4_naming_conventions/46_directory_names.md
Comment (niruiyu) [reply 4_naming_conventions/46_directory_names.md]: Update the question. This comment is similar to the one in below.
Usually ModuleName is short name of a Feature. E.g.: PCD(Dxe), Variable(RuntimeDxe), CpuMp(Pei).
Think about how a developer chooses a ModuleName when he creates a new module.

File: 4_naming_conventions/42_file_names.md
Comment (niruiyu) [reply 4_naming_conventions/42_file_names.md]: I still think this rule is not needed.
When the optional CpuArch and Vendor are absent in the file name, following rule can clearly describe this rule.

I want to simplify the rule so that people can remember/understand and follow that easily.

File: 4_naming_conventions/46_directory_names.md
Comment (niruiyu) [reply 4_naming_conventions/46_directory_names.md]: If you replace ModuleName with Feature, this rule is consistent to the rule defined for file names.

File: 4_naming_conventions/42_file_names.md
Comment (changab) [reply 4_naming_conventions/42_file_names.md]: We probably can mention "If the implementation supports >=2 or all CPU archs" case on the wiki page. That is ok we don't assign <arch> or <vendor> when the first time to implement the module. We won't rename the directory or INF. We will create a new INF file for arch or vendor instead.
Yeah, seems to me <ModuleName><Phase><CpuArch><Vendor>.inf covers <ModuleName><Phase>.inf case.

File: 4_naming_conventions/46_directory_names.md
Comment (changab) [reply 4_naming_conventions/46_directory_names.md]: yes. I can move directory names to prior file names

File: 4_naming_conventions/46_directory_names.md
Comment (changab) [reply 4_naming_conventions/46_directory_names.md]: Hmm, seems like we can replace all \<ModuleName\> with \<Feature\>. @mdkinney ?

File: 4_naming_conventions/42_file_names.md
Comment (changab) [reply 4_naming_conventions/42_file_names.md]: This section mainly talks about the source file.

File: 4_naming_conventions/42_directory_names.md
Comment (vlsunil) [inline 4_naming_conventions/42_directory_names.md]: In which case this will be required? Isn't generic module and arch specific folder inside sufficient always? For ex: Even if the module is created fresh for RISC-V, can we not name it generic since we don't know in future some other architecture may need it?

File: 4_naming_conventions/43_file_names.md
Comment (vlsunil) [inline 4_naming_conventions/43_file_names.md]: I still think this should be SmmAmdCpuFeaturesLib.inf. Since we can't change existing SmmCpuFeaturesLib.inf and CpuCommonFeaturesLib.inf, they can be considered as Intel related if different from AMD.

File: 4_naming_conventions/43_file_names.md
Comment (vlsunil) [inline 4_naming_conventions/43_file_names.md]: Is it Ia32 or IA32? Both are used in different places.

File: 4_naming_conventions/43_file_names.md
Comment (changab) [reply 4_naming_conventions/43_file_names.md]: What we define here is generic and mainly for the new module that is introduced to edk2
From the naming guidelines perspective, SmmAmdSmmCpuFeatureLib is compliant with <phase><CpuArch><Vendor><LibraryClassName><Dependency> file naming structure. So both SmmAmdSmmCpuFeaturesLib and SmmIntelSmmCpuFeaturesLib are correct.
However, SmmCpuFeaturesLib was implemented before the guidelines, the way to reconstruct the existing module is mentioned here, [The Guidelines of Reconstruct EDK II Modules for Processor Architectures and Vendors' Implementation](https://github.com/changab/tianocore.github.io/wiki/The-Guidelines-of-Reconstruct-EDK-II-Modules-for-Processor-Architectures-and-Vendors'-Implementation).
Take SmmCpuFeaturesLib as an example, we won't touch the existing INF file. However, the newly added INF file will follow the guidelines. I will remove SmmIntelSmmCpuFeaturesLib as I know that is confusing, but I will keep AMD one because that is a new file.

File: 4_naming_conventions/43_file_names.md
Comment (changab) [reply 4_naming_conventions/43_file_names.md]: Thanks for catching this. Should be Ia32

File: 4_naming_conventions/42_directory_names.md
Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: Those are all examples we can take for the module. Usually, if the module is only for the specific arch for sure then we can choose <Feature>Riscv64 (e,g, PlicDxeRiscV64). You can have CpuDxe/RiscV64 if you consider this module would have other archs. Or just use the generic name like CpuDxe. However, other archs may need a reconstruction later.  Conversations between contributor and maintainer may be required for the final decision.
Please also check the Wiki page, I added you as the collaborator,
[The Guidelines of Reconstruct EDK II Modules for Processor Architectures and Vendors' Implementation](https://github.com/changab/tianocore.github.io/wiki/The-Guidelines-of-Reconstruct-EDK-II-Modules-for-Processor-Architectures-and-Vendors'-Implementation)

Comment (vlsunil) [reply 4_naming_conventions/42_directory_names.md]: Okay. Thanks!

File: 4_naming_conventions/43_file_names.md
Comment (vlsunil) [reply 4_naming_conventions/43_file_names.md]: Okay.

File: 4_naming_conventions/43_file_names.md
Comment (vlsunil) [inline 4_naming_conventions/43_file_names.md]: Did you modify the directory path from IA32_X64 to Ia32_X64 in the examples above? For some reason, I still see IA32.

Comment (changab) [reply 4_naming_conventions/43_file_names.md]: Just pushed!

Comment (changab) [reply 4_naming_conventions/43_file_names.md]: Fix at commit bcb249b1788d7b11b182e2c2e1043fc7793a071a

File: 4_naming_conventions/42_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_directory_names.md]: Below rule
\<Feature\> \<Phase\> \[ \<CpuArch\> \[ \<Vendor\> \] \]
can cover below two rules:
\<Feature\> \<Phase\> and \<Feature\>\<Phase\>\<CpuArch\>\<Vendor\>

And it also restrict that Vendor only can be used when CpuArch exists.

File: 4_naming_conventions/42_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_directory_names.md]: PcdDxe driver can either be in Pcd/Dxe or PcdDxe folder.
I cannot see a clear reason why we use Pcd/Dxe instead of PcdDxe.
So, I think if we define the rule in this doc, Pcd/Dxe violates the rule. But that's ok IMO.

Comment (niruiyu) [reply 4_naming_conventions/42_directory_names.md]: What I mean is we could remove the rule "\<Feature\> / \<Phase\>".

File: 4_naming_conventions/42_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_directory_names.md]: We mixed the module path rule and the directory layout rule inside a module.
IMO, we needn't mention this rule here.

File: 4_naming_conventions/42_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_directory_names.md]: No need of this rule.

File: 4_naming_conventions/42_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_directory_names.md]: Similar concern: the rule after "or" mixed the module path rule and the directory layout rule inside a module.
We could remove the rule after "or".

File: 4_naming_conventions/42_directory_names.md
Comment (niruiyu) [inline 4_naming_conventions/42_directory_names.md]: Can we remove "_"?

File: 4_naming_conventions/43_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/43_file_names.md:143]: We'd better not define the rule for source files. Let's give freedom to developers. :)
Any change to the source file names only impacts the INF content.

File: 4_naming_conventions/43_file_names.md
Comment (niruiyu) [inline 4_naming_conventions/43_file_names.md]: The rule here mixed the directory rule and file name rule.
The directory rule is defined by chapter 4.2
So the file name rule here is:
\<Feature\>.inf
or
\<Feature\> \<CpuArch\> \<Vendor\>.inf

People may argue: why \<Feature\> should be part of file name because it's already part of directory path? Why \<Phase\> is not part of file name?

Can we align with Mike and propose a simple/clear/ONE rule for all?
I like the below rule you proposed:
\<Feature\> \<Phase\> \[ \<CpuArch\>\[\<Vendor\>\]\].inf

File: 4_naming_conventions/42_directory_names.md
Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: In order to have a consistent syntax in sections 4.2 and 4.3, I will use  "[]" for the optional element.

Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: will fix this in the next version.

File: 4_naming_conventions/42_directory_names.md
Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: This rule is defined specifically for the module which has no code shared between phases. PcdDxe follows the rule \<Feature\>\<Phase\>[\<CpuArch\>[\<Vendor\>]] which is different with \<Feature\>/\<Phase\>. But we can have \<Feature\>/\<Phase\>[/\<CpuArch\>[/\<Vendor\>]] for both \<Feature\>/\<Phase\> and \<Feature\>/\<Phase\>/\<CpuArch\>/\<Vendor\>.

File: 4_naming_conventions/42_directory_names.md
Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: Same as the above answer, this is a way to centralize different phase supports in one directory.
Pcd/Dxe/ is for the module that doesn't share code with PEI phase. Additionally, apply CpuArch>/<Vendor> on it.
Or do you prefer to have something like PcdDxe and PcdPei, no need Pcd/ directory for the different phases?

File: 4_naming_conventions/42_directory_names.md
Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: I am ok with this.

File: 4_naming_conventions/42_directory_names.md
Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: @niruiyu what do you mean here? Those two are the possible directories for the library module.

   - SmmIa32X64AmdSmmCpuFeaturesLib/
   - SmmArmAArch64SmmCpuFeaturesLib/
   - BaseMpInitLib/RiscV64/ 
                  /Ia32X64/ 
                  /ArmAArch64/  
                  /LoongArch64/
                  /

File: 4_naming_conventions/42_directory_names.md
Comment (changab) [reply 4_naming_conventions/42_directory_names.md]: > What I mean is we could remove the rule "<Feature> / <Phase>".

Or do you prefer to have something like PcdDxe and PcdPei, no need Pcd/ directory for the different phases?

File: 4_naming_conventions/43_file_names.md
Comment (changab) [reply 4_naming_conventions/43_file_names.md]: Phase is already mentioned in the directory (e.g. Pcd/__Dxe__) and we at least need [Feature] as the name of a file.

File: 4_naming_conventions/43_file_names.md
Comment (changab) [inline 4_naming_conventions/43_file_names.md]: @niruiyu
In order to have a consistent naming rule for \<CpuArch\>\<Vendor\>, I will apply [\<CpuArch\>[\<Vendor\>]] on this case as well. A long file name for \<CpuArch\>\<Vendor\> is expected.

File: 4_naming_conventions/43_file_names.md
Comment (changab) [reply 4_naming_conventions/43_file_names.md:143]: __FileName__ gives the freedom to developers when there is no \<CpuArch\> and \<Vendor\>, however, we should define the rules when considering the archs and vendors in the source file name. But I can remove the second part because it mixes up the file and directory naming.

File: 4_naming_conventions/43_file_names.md
Comment (changab) [reply 4_naming_conventions/43_file_names.md]: > @niruiyu
> In order to have a consistent naming rule for <CpuArch><Vendor>, I will apply [<CpuArch>[<Vendor>]] on this case as well. A long file name for <CpuArch><Vendor> is expected.

@niruiyu I still use [\<CpuArch\>][\<Vendor\>], so Vendor is unnecessary to have a dependence on CpuArch. The reason is the module is expected for X86 only, that is no need to have Ia32X64 for the vendor INF under X86 module. This concept is applied to the entire File name section. Please check the next version and see what else we can improve.

Comment (changab) [reply 4_naming_conventions/43_file_names.md]: > Please check the next version and see what else we can improve.

PR is updated.

**Pull request closed without merging**
